Add MPTs to xrpl.org docs (#2908)
XRPLF/xrpl-dev-portal: docs/references/http-websocket-apis/public-api-methods/path-and-order-book-methods/mpt_holders.md

---
blurb: Get the holders for a given `MPTokenIssuanceID` and ledger sequence.
labels:
  - Accounts
  - XRP
---

# mpt_holders

{% partial file="/docs/_snippets/mpts-disclaimer.md" /%}

For a given `MPTokenIssuanceID` and ledger sequence, `mpt_holders` returns all holders of that MPT and their balance. This method likely returns very large data sets, so you should expect to implement paging via the `marker` field. This API is only available using Clio, not rippled.

## Request Format

*Websocket*

```json
{
  "command": "mpt_holders",

>   "mpt_id": "00070C4495F14B0E44F78A264E41713C64B5F89242540EE255534400000000000000",

Reviewer: ```suggestion
  "mpt_issuance_id": "00070C4495F14B0E44F78A264E41713C64B5F89242540EE255534400000000000000",
```

  "ledger_index": "validated"
}
```

*JSON-RPC*

```json
{
  "method": "mpt_holders",
  "params": [
    {

>       "mpt_id": "00070C4495F14B0E44F78A264E41713C64B5F89242540EE255534400000000000000",

Reviewer: ```suggestion
      "mpt_issuance_id": "00070C4495F14B0E44F78A264E41713C64B5F89242540EE255534400000000000000",
```

      "ledger_index": "validated"
    }
  ]
}
```

The request contains the following parameters:

| Field             | Type                 | Required? | Description |
|:------------------|:---------------------|:----------|-------------|
| `mpt_issuance_id` | string               | Yes       | The `MPTokenIssuance` to query. |
| `ledger_index`    | string or number (positive integer) | No | The ledger index of the max ledger to use, ora shortcut string to choose a ledger automatically. You must specify either ledger_index or ledger_hash. |

> | `ledger_hash`     | string               | No        | A 20-byte hex string for the ma ledger version to use. You must specify either ledger_index or ledger_hash. |

Reviewer 2: ```suggestion
| `ledger_hash`     | string               | No        | A 32-byte hex string for the ledger version to use. You must specify either ledger_index or ledger_hash. |
```

Author: I see ledger_hash defined as a 20-byte hex string in several places in the docs. Why would this instance be different?

e.g.
https://xrpl.org/docs/references/http-websocket-apis/public-api-methods/ledger-methods/ledger_data#request-format
https://xrpl.org/docs/references/http-websocket-apis/public-api-methods/ledger-methods/ledger_closed#response-format

Reviewer 2: I don't know. Possibly the wrong one was copied and reused.

Author: That's a bigger problem, then. Thanks!

| `marker`          | string               | No        | Used to continue your query where it left off in paginating. |
| `limit`           | number (positive integer) | No   | Specify a limit to the number of MPTs returned. |

## Response Format

```json
{
    "mpt_issuance_id": "000004C463C52827307480341125DA0577DEFC38405B0E3E",
    "limit":50,
    "ledger_index": 2,
    "mptokens": [{
        "account": "rEiNkzogdHEzUxPfsri5XSMqtXUixf2Yx",
        "flags": 0,
        "mpt_amount": "20",
        "mptoken_index": "36D91DEE5EFE4A93119A8B84C944A528F2B444329F3846E49FE921040DE17E65"
    },
    {
        "account": "rrnAZCqMahreZrKMcZU3t2DZ6yUndT4ubN",
        "flags": 0,
        "mpt_amount": "1",
        "mptoken_index": "D137F2E5A5767A06CB7A8F060ADE442A30CFF95028E1AF4B8767E3A56877205A"
    }],
    "validated": true
}
```

### Response Fields

> The response follows the [standard format](https://xrpl.org/docs/references/http-websocket-apis/api-conventions/response-formatting/), with the result containing the following fields:

Reviewer 2: ```suggestion
The response follows the [standard format][], with the result containing the following fields:
```

      "Fee": "10",
      "Flags": 64,
      "Sequence": 303,
      "SigningPubKey": "ED39955DEA2D083C6CBE459951A0A84DB337925389ACA057645EE6E6BA99D4B2AE",
      "TransactionType": "MPTokenIssuanceCreate",
      "TxnSignature": "80D7B7409980BE9854F7217BB8E836C8A2A191E766F24B5EF2EA7609E1420AABE6A1FDB3038468679081A45563B4D0B49C08F4F70F64E41B578F288A208E4206",
      "ctid": "C000013100000000",
      "date": 760643692,
      "hash": "E563D7942E3E4A79AD73EC12E9E4C44B7C9950DF7BF5FDB75FAD0F5CE0554DB3",
      "inLedger": 305,
      "ledger_index": 305,
      "meta": {
         "AffectedNodes": [...],
         "TransactionIndex": 0,
         "TransactionResult": "tesSUCCESS",
         "mpt_issuance_id": "0000012F72A341F09A988CDAEA4FF5BE31F25B402C550ABE"
      },
      "status": "success",
      "validated": true
   }
}
```

##### Object
An `mpt_issuance_id` field is provided in JSON MPTokenIssuance objects (not available for binary). The following APIs are impacted: `ledger_data` and `account_objects`.

##### Example
Example of an `account_objects` response:

```json
{
   "result": {
      "account": "rBT9cUqK6UvpvZhPFNQ2qpUTin8rDokBeL",
      "account_objects": [
      {

>       "AssetScale": 2,

Reviewer 2: ```suggestion
          {
              "AssetScale": 2,
```

---
blurb: Get the holders for a given `MPTokenIssuanceID` and ledger sequence.
labels:
  - Accounts
  - XRP
---

> # mpt_holders

Reviewer: should mention that this API is only available in clio, not rippled

| Field                  | Type    | Description                               |
|:-----------------------|:--------|:------------------------------------------|
| `mpt_issuance_id`      | string  | The `MPTokenIssuance` queried             |
| `mptokens`             | array   | An array of mptokens. Includes all relevant fields in the underlying MPToken object. |
| `marker`               | string  | Used to continue querying where we left off when paginating. Omitted if there are no more entries after this result. |
| `limit`                | number  | The limit, as specfied in the request
| `ledger_index`         | number  | The index of the ledger used. |

An `mptoken` object has the following parameters:

| Field                  | Type    | Description |
|:-----------------------|:--------|:------------------------------------------|
| `account`              | string  | The account address of the holder who owns the `MPToken`. |
| `flags`                | number  | The flags assigned to the`MPToken` object. |
| `mpt_amount`           | string  | Base 10-encoded amount of the holder's balance. |

> | `locked_amount`        | string  | Base 10-encoded amount of the locked balance. (Can be omitted if the value is 0.) |

Reviewer: `locked_amount`/`LockedAmount` is now removed everywhere

| `mptoken_index`        | string  | Key of the `MPToken` object. |

#### Synthetic mpt_issuance_id field
`MPTokenIssuanceID` is an identifier that allows you to specify an `MPTokenIssuance` in RPCs. The server adds a synthetically parsed `mpt_issuance_id` field to API responses to avoid the need for client-side parsing of the `MPTokenIssuanceID`.

##### Transaction Metadata
An `mpt_issuance_id` field is provided in JSON transaction metadata (not available for binary) for all successful `MPTokenIssuanceCreate` transactions. The following APIs are impacted: `tx`, `account_tx`, `subscribe` and `ledger`.

> 

Reviewer: rest of the file addition doesn't feel belong in this file. [This](https://xrpl.org/docs/references/protocol/transactions/metadata) is a better place for them

      "TransactionType": "MPTokenIssuanceCreate",
      "TxnSignature": "80D7B7409980BE9854F7217BB8E836C8A2A191E766F24B5EF2EA7609E1420AABE6A1FDB3038468679081A45563B4D0B49C08F4F70F64E41B578F288A208E4206",
      "ctid": "C000013100000000",
      "date": 760643692,
      "hash": "E563D7942E3E4A79AD73EC12E9E4C44B7C9950DF7BF5FDB75FAD0F5CE0554DB3",
      "inLedger": 305,
      "ledger_index": 305,
      "meta": {
         "AffectedNodes": [...],
         "TransactionIndex": 0,
         "TransactionResult": "tesSUCCESS",
         "mpt_issuance_id": "0000012F72A341F09A988CDAEA4FF5BE31F25B402C550ABE"
      },
      "status": "success",
      "validated": true
   }
}
```

##### Object
An `mpt_issuance_id` field is provided in JSON MPTokenIssuance objects (not available for binary). The following APIs are impacted: `ledger_data` and `account_objects`.

##### Example
Example of an `account_objects` response:

```json
{
   "result": {
      "account": "rBT9cUqK6UvpvZhPFNQ2qpUTin8rDokBeL",
      "account_objects": [
      {
      "AssetScale": 2,
            "Flags": 64,
            "Issuer": "rBT9cUqK6UvpvZhPFNQ2qpUTin8rDokBeL",
            "LedgerEntryType": "MPTokenIssuance",

>             "OutstandingAmount": "5a",

Reviewer: ```suggestion
            "OutstandingAmount": "100",
```

We need to change to a number that's base 10 instead of hex
